### PR TITLE
Minor gc improvements

### PR DESCRIPTION
**Pull request by wizzardo**

### CODE REVIEW

File: bt-core/src/main/java/bt/torrent/messaging/Assignment.java [hunk 1]
```diff
@@ -72,6 +72,13 @@ ConnectionKey getConnectionKey() {
     private void claimPiecesIfNeeded() {
         if (pieces.size() < MAX_SIMULTANEOUSLY_ASSIGNED_PIECES) {
             Bitfield peerBitfield = pieceStatistics.getPeerBitfield(connectionKey).get();
+            if(!assignments.isEndgame()){
```
Comment (atomashpolskiy) [inline bt-core/src/main/java/bt/torrent/messaging/Assignment.java:75]: So this is same as below, but without shuffling?

**Review by atomashpolskiy — COMMENTED** [review]
All good, just one small suggestion (a more descriptive exception) and one question

File: bt-core/src/main/java/bt/torrent/messaging/Assignment.java (hunk 1 above)
Comment (wizzardo) [reply bt-core/src/main/java/bt/torrent/messaging/Assignment.java:75]: it's the same as the code below, shuffling applied only for 'end-game' as I understand, the point of this change is to read only up to 3 values from the stream, not the whole stream and place it into array

File: bt-core/src/main/java/bt/protocol/Protocols.java [hunk 2]
```diff
@@ -157,17 +157,29 @@ public static Integer readInt(ByteBuffer buffer) {
...
+     * @return Decoded value
+     * @throws NullPointerException if there are insufficient bytes in buffer
      *         (i.e. <b>buffer.remaining()</b> &lt; 4)
      * @since 1.9
      */
-    public static Integer readInt(ByteBufferView buffer) {
+    public static int readInt(ByteBufferView buffer) {
         if (buffer.remaining() < Integer.BYTES) {
-            return null;
+            throw new NullPointerException();
```
[the remark this answers is not in the document]
Comment (wizzardo) [reply bt-core/src/main/java/bt/protocol/Protocols.java:167]: it was just a NullPointerException before, but ok let's have a bit more info here

**Pull request merged**